src/EditorFeatures/CSharpTest/CompleteStatement/CSharpCompleteStatementCommandHandlerTests.cs

```diff
@@ -0,0 +1,2577 @@
+@"
+class C
+{
+    static void Main()
+    {
+        for ( int i = 0, int j = 0$$)
+        int j;
+";
+
+            VerifyNoSpecialSemicolonHandling(code);
+        }
+
+        [WpfFact, Trait(Traits.Feature, Traits.Features.CompleteStatement)]
+        public void ForLoopNewInInitializer1()
+        {
+            var code =
+@"
+class C
+{
+    static void Main(string[] args)
+    {
+        for (C1 i = new C1($$))
+        int j;
+    }
+}
+public class C1
+{
+    public static C1 operator ++(C1 obj)
+    {
+        return obj;
+    }
+}
+";
+
+            var expected =
+@"
+class C
+{
+    static void Main(string[] args)
+    {
+        for (C1 i = new C1();$$)
+        int j;
+    }
+}
+public class C1
+{
+    public static C1 operator ++(C1 obj)
+    {
+        return obj;
+    }
+}
+";
+
+            VerifyTypingSemicolon(code, expected);
+        }
+
+        [WpfFact, Trait(Traits.Feature, Traits.Features.CompleteStatement)]
+        public void ForLoopNewInInitializer_MissingOneParen()
+        {
+            var code =
+@"
+class C
+{
+    static void Main(string[] args)
+    {
+        for (C1 i = new C1()$$
+        int j;
+    }
+}
+public class C1
+{
+    public static C1 operator ++(C1 obj)
+    {
+        return obj;
+    }
+}
+";
+
+            VerifyNoSpecialSemicolonHandling(code);
+        }
+
+        [WpfFact, Trait(Traits.Feature, Traits.Features.CompleteStatement)]
+        public void ForLoopNewInInitializer2_MissingBothParens()
+        {
+            // only adding one closing paren
+            var code =
+@"
+class C
+{
+    static void Main(string[] args)
+    {
+        for (C1 i = new C1($$
+        int j;
+    }
+}
+public class C1
+{
+    public static C1 operator ++(C1 obj)
+    {
+        return obj;
+    }
+}
+";
+
+            VerifyNoSpecialSemicolonHandling(code);
+        }
+
+        [WpfFact, Trait(Traits.Feature, Traits.Features.CompleteStatement)]
+        public void ForLoopInitializer()
+        {
+            var code =
+@"
+class C
+{
+    static void Main(string[] args)
+    {
+        string s = ""abcdefghij"";
+        for (int i = s.IndexOf(""bcd""$$) i < 10; i++)
+";
+
+            var expected =
+@"
+class C
+{
+    static void Main(string[] args)
+    {
+        string s = ""abcdefghij"";
+        for (int i = s.IndexOf(""bcd"");$$ i < 10; i++)
+";
+
+            VerifyTypingSemicolon(code, expected);
+        }
+
+        [WpfFact, Trait(Traits.Feature, Traits.Features.CompleteStatement)]
+        public void ForLoopInitializer_MissingParen()
+        {
+            var code =
+@"
+class C
+{
+    static void Main(string[] args)
+    {
+        string s = ""abcdefghij"";
+        for (int i = s.IndexOf(""bcd""$$ i < 10; i++)
+";
+
+            var expected =
+@"
+class C
+{
+    static void Main(string[] args)
+    {
+        string s = ""abcdefghij"";
+        for (int i = s.IndexOf(""bcd"";$$ i < 10; i++)
+";
+
+            VerifyTypingSemicolon(code, expected);
```

Comment:
>VerifyTypingSemicolon [](start = 12, length = 21)

Interesting, seems like we do not complete for a standalone invocation with missing parens, but if such an invocation with missing parens is within the for loop header, we do perform the completion. Any reason for this differing behavior?

Reply:
In a standalone invocation with missing parenthesis, we don't know where the end of the invocation is.  In a for loop header, if the first semicolon is present then we have a much better idea (not always perfect) where the parenthesis would go.